sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala

```diff
@@ -62,6 +66,122 @@ private[sql] class ResolveDataSource(sparkSession: SparkSession) extends Rule[Lo
 }
 
 /**
+ * Preprocess some DDL plans, e.g. [[CreateTable]], to do some normalization and checking.
+ */
+case class PreprocessDDL(conf: SQLConf) extends Rule[LogicalPlan] {
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    // When we CREATE TABLE without specifying the table schema, we should fail the query if
+    // bucketing information is specified, as we can't infer bucketing from data files currently,
+    // and we should ignore the partition columns if it's specified, as we will infer it later, at
+    // runtime.
+    case c @ CreateTable(tableDesc, _, None) if tableDesc.schema.isEmpty =>
+      if (tableDesc.bucketSpec.isDefined) {
+        failAnalysis("Cannot specify bucketing information if the table schema is not specified " +
+          "when creating and will be inferred at runtime")
+      }
+
+      val partitionColumnNames = tableDesc.partitionColumnNames
+      if (partitionColumnNames.nonEmpty) {
+        // The table does not have a specified schema, which means that the schema will be inferred
+        // at runtime. So, we are not expecting partition columns and we will discover partitions
+        // at runtime. However, if there are specified partition columns, we simply ignore them and
+        // provide a warning message.
+        logWarning(
+          s"Specified partition columns (${partitionColumnNames.mkString(",")}) will be " +
+            s"ignored. The schema and partition columns of table ${tableDesc.identifier} will " +
+            "be inferred.")
+        c.copy(tableDesc = tableDesc.copy(partitionColumnNames = Nil))
+      } else {
+        c
+      }
+
+    // Here we normalize partition, bucket and sort column names, w.r.t. the case sensitivity
+    // config, and do various checks:
+    //   * column names in table definition can't be duplicated.
+    //   * partition, bucket and sort column names must exist in table definition.
+    //   * partition, bucket and sort column names can't be duplicated.
+    //   * can't use all table columns as partition columns.
+    //   * partition columns' type must be AtomicType.
+    //   * sort columns' type must be orderable.
```

Comment:
because hive CREATE TABLE syntax is special, it will never hit this exception `partition column names must exist in table definition.` Instead, hive may specify partition column names that exist in table definition, and will hit `column names in table definition can't be duplicated`.